hal/src/dmac/channel.rs

```diff
@@ -134,15 +136,22 @@ impl<Id: ChId, S: Status> Channel<Id, S> {
     /// If an interrupt were to change the CHID register, we would be faced
     /// with undefined behaviour.
     #[cfg(any(feature = "samd11", feature = "samd21"))]
-    fn with_chid<F: Fn(&DMAC)>(&mut self, dmac: &DMAC, fun: F) {
+    fn with_chid<F: FnMut(&DMAC)>(&mut self, mut fun: F) {
         cortex_m::interrupt::free(|_| {
-            // SAFETY: this is actually safe as long as we write a correct channel number to
-            // the CHID register
-            unsafe {
+            // SAFETY: This is ONLY safe if the individual channels are GUARANTEED not to
+            // mess with either:
+            // - The global DMAC configuration
+            // - The configuration of other channels.
+            //
+            // In practice, this means that the DMAC registers should only be accessed
+            // through the `with_chid` method.
+            let dmac = unsafe {
+                let dmac = Peripherals::steal().DMAC;
```

Comment:
This is a somewhat ad-hoc approach. I used to do something like this in the gpio module. I came up with and alternative that I think is better though. You can create your own PAC-like, ZST, smart pointers that provide an API for each channel's registers. It's basically as if you split the PAC struct into segments according to channel. Then each channel can take ownership of that ZST smart pointer. I think that ends up being cleaner than stealing.

See the gpio module for an example. It's a little bit more complicated there, because I needed to implement it for both Pin and DynPin, but I think the documentation on the RegisterInterface trait should explain it well.

Reply:
Also check out some of the clock modules in my clock_rebase branch. It's linked in the clock RFC issue. I used this pattern in almost every module of clock::v2.